### ISSUE TITLE
Linkerd otel collector exposes port 4318(http)

### ISSUE THREAD

**Issue by uralsemih** [linkerd/linkerd2#9366]
### What problem are you trying to solve?

The collector exposes port 4318(http) using OTLP.

### How should the problem be solved?

OTLP/HTTP port 4318 needs to be added collector.

### Any alternatives you've considered?

I tried to add 4318 manually in deployment and service but I could not able to access traces where I am getting 400 Bad Request Error.

Deployment
```
apiVersion: apps/v1
kind: Deployment
metadata:
  annotations:
    deployment.kubernetes.io/revision: "1"
    kubectl.kubernetes.io/last-applied-configuration: |
  creationTimestamp: "2022-09-10T19:44:11Z"
  generation: 1
  labels:
    app.kubernetes.io/name: collector
    app.kubernetes.io/part-of: Linkerd
    component: collector
    linkerd.io/extension: jaeger
  name: collector
  namespace: linkerd-jaeger
  resourceVersion: "94872"
  uid: ab1c7da8-0d5c-4736-afd3-9d2139cf36a0
spec:
  minReadySeconds: 5
  progressDeadlineSeconds: 120
  replicas: 1
  revisionHistoryLimit: 10
  selector:
    matchLabels:
      component: collector
  strategy:
    rollingUpdate:
      maxSurge: 25%
      maxUnavailable: 25%
    type: RollingUpdate
  template:
    metadata:
      annotations:
        cluster-autoscaler.kubernetes.io/safe-to-evict: "true"
        config.linkerd.io/proxy-await: enabled
        linkerd.io/inject: enabled
        prometheus.io/path: /metrics
        prometheus.io/port: "8888"
        prometheus.io/scrape: "true"
      creationTimestamp: null
      labels:
        component: collector
        linkerd.io/extension: jaeger
    spec:
      containers:
      - command:
        - /otelcol
        - --config=/conf/collector-config.yaml
        env:
        - name: GOGC
          value: "80"
        - name: POD_NAMESPACE
          valueFrom:
            fieldRef:
              apiVersion: v1
              fieldPath: metadata.namespace
        image: otel/opentelemetry-collector:0.43.0
        imagePullPolicy: Always
        name: ot-collector
        ports:
        - containerPort: 13133
          protocol: TCP
        - containerPort: 4317
          protocol: TCP
        - containerPort: 4318
          protocol: TCP
        - containerPort: 55678
          protocol: TCP
        - containerPort: 9411
          protocol: TCP
        - containerPort: 14268
          protocol: TCP
        - containerPort: 14250
          protocol: TCP
        - containerPort: 8888
          protocol: TCP
```
Service
```
apiVersion: v1
kind: Service
metadata:
  annotations:
    kubectl.kubernetes.io/last-applied-configuration: |
      {"apiVersion":"v1","kind":"Service","metadata":{"annotations":{},"labels":{"component":"collector","linkerd.io/extension":"jaeger"},"name":"collector","namespace":"linkerd-jaeger"},"spec":{"ports":[{"name":"otlp","port":4317,"protocol":"TCP","targetPort":4317},{"name":"http","port":4318,"protocol":"TCP","targetPort":4318},{"name":"opencensus","port":55678,"protocol":"TCP","targetPort":55678},{"name":"zipkin","port":9411,"protocol":"TCP","targetPort":9411},{"name":"jaeger-thrift","port":14268,"protocol":"TCP","targetPort":14268},{"name":"jaeger-grpc","port":14250,"protocol":"TCP","targetPort":14250}],"selector":{"component":"collector"},"type":"ClusterIP"}}
  creationTimestamp: "2022-09-10T19:44:11Z"
  labels:
    component: collector
    linkerd.io/extension: jaeger
  name: collector
  namespace: linkerd-jaeger
  resourceVersion: "97126"
  uid: f1dcc7f6-3fde-4f96-a202-d048100fb946
spec:
  clusterIP: 10.43.38.190
  clusterIPs:
  - 10.43.38.190
  internalTrafficPolicy: Cluster
  ipFamilies:
  - IPv4
  ipFamilyPolicy: SingleStack
  ports:
  - name: otlp
    port: 4317
    protocol: TCP
    targetPort: 4317
  - name: http
    port: 4318
    protocol: TCP
    targetPort: 4318
  - name: opencensus
    port: 55678
    protocol: TCP
    targetPort: 55678
  - name: zipkin
    port: 9411
    protocol: TCP
    targetPort: 9411
  - name: jaeger-thrift
    port: 14268
    protocol: TCP
    targetPort: 14268
  - name: jaeger-grpc
    port: 14250
    protocol: TCP
    targetPort: 14250
  selector:
    component: collector
  sessionAffinity: None
  type: ClusterIP
status:
  loadBalancer: {}
```

```
Ingress
apiVersion: networking.k8s.io/v1
kind: Ingress
metadata:
  name: otel-collector
  namespace: linkerd-jaeger
  annotations:
    nginx.ingress.kubernetes.io/service-upstream: "true"
    nginx.ingress.kubernetes.io/enable-cors: "true"
    nginx.ingress.kubernetes.io/cors-allow-origin: "http://demo.sre.localhost"
    nginx.ingress.kubernetes.io/cors-allow-methods: "PUT, GET, POST, OPTIONS, DELETE"
    nginx.ingress.kubernetes.io/cors-allow-headers: "DNT,X-CustomHeader,X-LANG,Keep-Alive,User-Agent,X-Requested-With,If-Modified-Since,Cache-Control,Content-Type,X-Api-Key,X-Device-Id,Access-Control-Allow-Origin,b3,Authorization"
    nginx.ingress.kubernetes.io/configuration-snippet: |
      proxy_hide_header l5d-remote-ip;
      proxy_hide_header l5d-server-id;
spec:
  ingressClassName: nginx
  rules:
    - host: otel-collector.localhost
      http:
        paths:
          - path: /v1/traces
            pathType: Prefix
            backend:
              service:
                name: collector
                port:
                  number: 4318
```

**Comment by uralsemih** [comment]
I created this PR to fix this
https://github.com/linkerd/linkerd2/pull/9379

**Comment by olix0r** [comment]
@uralsemih Thanks for submitting a PR.

> ### What problem are you trying to solve?
> 
> The collector exposes port 4318(http) using OTLP.

Can you elaborate on what problem you're hitting here? Clearly nothing in Linkerd needs this port to be exposed. What does?

The `containerPorts` in a pod manifest are purely informational and should have no bearing on whether a port is exposed. Is this change just for completeness? Or what problems were you hitting?

**Comment by uralsemih** [comment]
@olix0r due to React can not use grpc with opentelemetry where it needs to use 4318(http) for traces, I am aware that the collector deployed with linkerd-jaeger using a bit old version(0.43.0) which seems do not have 4318. As well as, I added this port in deployment and service.yaml cause in my case I should expose the otel-collector via ingress using this port to send traces from browser.

If you check the opentelemetry collector the default otlp/http port is 4318 so its valuable if we can add this in the collector that linkerd provided.
https://opentelemetry.io/docs/reference/specification/protocol/otlp/

**Comment by adleong** [comment]
Hi @uralsemih.  I can confirm that the collector does serve OTLP HTTP on port 4318.  However, I have the same questions as @olix0r: I would expect that React should be able to send traces to the collector on port 4318 even if it is not included in the container or service spec.  What problems are you running into?

**Comment by uralsemih** [comment]
Hi @adleong,
The existing collector that installed via linkerd-jaeger has not 4318 port cause it is using older version and port did not added.. I tried to expose otel-collector via ingress using port `4318` and `/v1/traces` to send traces from browser but I could not able to till I ugpraded the version of otel collector and added this port in container.

**Comment by adleong** [comment]
Looking at the collector logs for linkerd stable-2.12.0 I see:

```
2022-09-15T22:35:14.561Z        info    otlpreceiver/otlp.go:87 Starting HTTP server on endpoint 0.0.0.0:4318   {"kind": "receiver", "name": "otlp"}
```

**Comment by uralsemih** [comment]
Did you try to expose `otel-collector` via port-forward with 4318 ? It is not accessible

**Comment by adleong** [comment]
Yes, and at the very least it serves HTTP:

```
curl -v localhost:4318          
*   Trying 127.0.0.1:4318...
* Connected to localhost (127.0.0.1) port 4318 (#0)
> GET / HTTP/1.1
> Host: localhost:4318
> User-Agent: curl/7.74.0
> Accept: */*
> 
* Mark bundle as not supporting multiuse
< HTTP/1.1 404 Not Found
< Content-Type: text/plain; charset=utf-8
< X-Content-Type-Options: nosniff
< Date: Fri, 16 Sep 2022 21:18:54 GMT
< Content-Length: 19
< 
404 page not found
* Connection #0 to host localhost left intact
```

**Comment by uralsemih** [comment]
Port might be open but I could not able to send traces through `4318` and` /v1/traces`. Port-forward and try to access at least from browser `http://localhost:4318/v1/traces`. You should get a 404 error rather than 405. And try to upgrade otel-collector image with adding 4318 port in deployment, after you can able to access

**Comment by uralsemih** [comment]
@adleong Did you have any chance to test?

**Comment by adleong** [comment]
@uralsemih to be honest, I'm having a bit of trouble following along here.  Initially this issue was about adding container ports definitions to the collector deployment spec but it seems like maybe now it's about upgrading the collector image version.  Would you mind updating the issue description with a clear explanation of the problem, suggested solution, and how to test?

**Comment by uralsemih** [comment]
@adleong Because there is a relation between each other.

Without updating the image version, adding a port in deployment/service

- kubectl port-forward collector -n linkerd-jaeger
- Try to send traces from the browser via http://localhost:4318/v1/traces
- Will get a 404 error

Update the otel image version from 0.43 to 0.59, adding 4318 port in deployment as a container port

- kubectl port-forward collector -n linkerd-jaeger
- Try to send traces from the browser via http://localhost:4318/v1/traces
- Will get 405, where it is accesible now

**Comment by kleimkuhler** [comment]
Closed by #9379
